packages/@aws-cdk/aws-appmesh/lib/private/utils.ts

```diff
@@ -95,13 +97,42 @@ export function renderMeshOwner(resourceAccount: string, meshAccount: string) :
 }
 
 /**
- * This is the helper method to validate the length of match array when it is specified.
+ * This is the helper method to validate the length of HTTP match array when it is specified.
  */
-export function validateMatchArrayLength(headers?: HeaderMatch[]) {
+export function validateHttpMatchArrayLength(headers?: HeaderMatch[], queryParameter?: QueryParameterMatch[]) {
   const MIN_LENGTH = 1;
   const MAX_LENGTH = 10;
 
   if (headers && (headers.length < MIN_LENGTH || headers.length > MAX_LENGTH)) {
     throw new Error(`Number of headers provided for matching must be between ${MIN_LENGTH} and ${MAX_LENGTH}, got: ${headers.length}`);
   }
+
+  if (queryParameter && (queryParameter.length < MIN_LENGTH || queryParameter.length > MAX_LENGTH)) {
+    throw new Error(`Number of query parameters provided for matching must be between ${MIN_LENGTH} and ${MAX_LENGTH}, got: ${queryParameter.length}`);
+  }
+}
+
+/**
+ * This is the helper method to validate the length of gRPC match array when it is specified.
+ */
+export function validateGrpcMatchArrayLength(metadata?: HeaderMatch[]) {
+  const MIN_LENGTH = 1;
+  const MAX_LENGTH = 10;
+
+  if (metadata && (metadata.length < MIN_LENGTH || metadata.length > MAX_LENGTH)) {
+    throw new Error(`Number of metadata provided for matching must be between ${MIN_LENGTH} and ${MAX_LENGTH}, got: ${metadata.length}`);
+  }
+}
+
+/**
+ * This is the helper method to validate at least one of gRPC match type is defined.
+ */
+export function validateGprcMatch(match: GrpcRouteMatch) {
+  for (const property in match) {
```

Comment:
Unfortunately, I don't think this does what you want it to do.

TypeScript uses a structural type system, so for example, you could do:

```ts
const myMatch = {
  randomProperty: true,
};
```

and then pass `myMatch` to where `GrpcRouteMatch` is expected, and your method would return `true`.

I think you might have to spell out the condition explicitly, something like `return match.serviceName == undefined && match.methodName === undefined && match.metadata === undefined`.

Reply:
I am not sure if I am understanding your concern here.

I tried above example and when `myMatch` is inserted, I am seeing the compile-time error, "type '{ randomProperty: boolean; }' has no properties in common with type 'GrpcRouteMatch'."

Reply:
I think I understood the issue here. Looks like if one of the property of the `match` matches with `GrpcRouteMatch` then there would no error be shown.

```ts
const myMatch = {
      serviceName: 'test',  // This exists in GrpcRouteMatch
      hostname: 'GatewayRouteHostnameMatch',  // This does NOT exist in GrpcRouteMatch
    };

    validateGrpcMatch(myMatch);
```

However, would this actually a problem? Since the `validateGrpcMatch` function is private and thus as long as the developer uses it with correct context I feel like it is ok.

Reply:
It doesn't matter that the function is private - it gets called with a value provided by the customer of the AppMesh module.

Reply:
@skinny85 - Ok. I think I am understanding the the problem here. However, I still think `return match.serviceName == undefined && match.methodName === undefined && match.metadata === undefined` is unnecessary because the criteria for `validateGrpcMatch` is to validate if "at least one match property is defined" and if `myMatch` example does not contain matching property then then there will be a compile-time error.

Reply:
Still an issue.

Reply:
To add a note, if we are doing the explicit check, I think it should check whether they are not empty rather than checking if they not `undefined`.

Reply:
Concluded to perform explicit check to ensure passed object contains necessary properties.